src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+      val pairedEnd: Boolean                  = readStructures.count(_.template.nonEmpty) > 1
+
+      val records = fastqRecords.zipWithIndex.flatMap { case (record, recordIndex) =>
+        val readStructure = templateReadStructures(recordIndex)
+        if (readStructure.isEmpty) {
+          None
+        }
+        else {
+          if (1 < recordIndex) throw new IllegalArgumentException("Found more than two ends with template bases")
+          val samRecord: SAMRecord = makeSAMRecord(header, record, readStructure, sampleId, pairedEnd, recordIndex == 0)
+          
+          // set attributes: sample and molecular barcodes
+          samRecord.setAttribute("BC", sampleBarcodeReadBases.mkString(SampleBarcodeDelimiter))
+          val umiValue = molecularBarcodes.mkString(SampleBarcodeDelimiter)
+          if (umiValue.nonEmpty) samRecord.setAttribute(umiTag, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+
+          if (multiUmiTags.nonEmpty) {
+            val numMolecularBarcodeSegments = getNumMolecularBarcodes(fastqRecords)
+            if (numMolecularBarcodeSegments == multiUmiTags.length) {
+              multiUmiTags.zip(molecularBarcodes).foreach { case (tag, barcode) => samRecord.setAttribute(tag, barcode) }
+            }
+            else if (multiUmiTags.length == 1) {
+              samRecord.setAttribute(multiUmiTags.head, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+            }
+            else { // more than one tag, but not the same as the number of molecular barcodes
+              unreachable(s"Did not find only one SAM tag or the same # of SAM tags for the molecular barcodes (${multiUmiTags.length} != $numMolecularBarcodeSegments)")
+            }
+          }
+          Some(samRecord)
+        }
+      }
+      AssignedRecords(sampleOrdinal=sampleOrdinal, records=records)
+    }
+
+    def getBarcodeMetrics: Iterable[SampleBarcodeMetric] = {
+      SampleBarcodeMetric.finalizeMetrics(this.barcodeToMetrics, this.noMatchBarcode)
+      this.barcodeToMetrics.values
+    }
+
+    private def noMatchBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics(noMatchBarcode)
+
+    /** Gets the # of mismatches for the given sample and observed sample barcode read bases. */
+    private def sampleOrdinalAndMismatches(sample: Sample, sampleBarcodeReadBases: String): (Int, Int) = {
+      val sampleBarcode = sample.sampleBarcode match {
+        case Some(barcode) => barcode
+        case None => throw new IllegalArgumentException(s"Sample barcode required for sample with id '${sample.sampleId}'")
+      }
+      val forwardMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.barcodeBytes)
+      val numMismatches = if (matchReverseComplement) {
+        if (!sampleBarcode.isInstanceOf[CustomSampleBarcode]) unreachable("Cannot match the reverse complement unless the barcode is of type 'CustomSampleBarcode'.")
```

Comment:
1) Is this truly unreachable?
2) The error message would be meaningless to a user 
3) Why limit it to `CustomSampleBarcode` only?